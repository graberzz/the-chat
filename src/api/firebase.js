import firebase from "firebase";

const config = {
  apiKey: "AIzaSyBnZTs_h_o0Uw5m706dcTbP54VtSM5ZZYE",
  authDomain: "chat-8a443.firebaseapp.com",
  databaseURL: "https://chat-8a443.firebaseio.com",
  projectId: "chat-8a443",
  storageBucket: "chat-8a443.appspot.com",
  messagingSenderId: "660697500625"
};

if (firebase.apps.length === 0) {
  firebase.initializeApp(config);
}

const database = firebase.database();

export { firebase };

export default database;
