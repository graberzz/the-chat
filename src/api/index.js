import db, { firebase } from './firebase';

const messagesRef = db.ref('messages');
const typersRef = db.ref('typing');
const roomsRef = db.ref('rooms');

const auth = firebase.auth();
auth.useDeviceLanguage();
const provider = new firebase.auth.GoogleAuthProvider();

const getUser = googleUser => ({
  username: googleUser.displayName,
  email: googleUser.email,
  avatar: googleUser.photoURL
});

export const onAuthChange = (onSignIn, onSignOut) => {
  auth.onAuthStateChanged(user => {
    if (user) {
      onSignIn(getUser(user));
    } else {
      onSignOut();
    }
  });
};

export const login = () =>
  new Promise((res, rej) => {
    auth
      .signInWithPopup(provider)
      .then(({ user }) => {
        res(getUser(user));
      })
      .catch(rej);
  });

export const logout = () => {
  return auth.signOut();
};

export const sendMessage = (user, message, roomId, onError) => {
  messagesRef.push().set(
    {
      user,
      message,
      roomId,
      timestamp: Date.now(),
    }, onError);
};

export const addRoom = (name, creator = 'Cthulhu', onError) => {
  roomsRef.push().set({
    name,
    creator,
  }, onError);
}

export const watchRooms = (callback) => {
  roomsRef.on('child_added', (data) => callback({
    ...data.val(),
    id: data.key,
  }));
}

export const watchNewMessages = (timestamp, callback) => {
  messagesRef.on('child_added', (data) => {
    const msg = data.val();

    // skip messages older than timestamp
    if (msg.timestamp < timestamp) return;

    callback({
      ...msg,
      id: data.key,
    });
  });
};

export const getRoomMessages = (roomId) => new Promise((res, rej) => {
  messagesRef
    .orderByChild('roomId')
    .equalTo(roomId)
    .once('value', (snap) => res(snap.val() || {}));
});

export const setTyping = (username, roomId, unsetTimeout) => {
  typersRef.child(roomId).child(username).set({
    username
  });

  return setTimeout(() => unsetTyping(username, roomId), unsetTimeout);
};

export const unsetTyping = (username, roomId) => {
  typersRef.child(roomId).child(username).remove();
};

export const watchTypers = (callback) => {
  typersRef.on('value', (snap) => {
    callback(snap.val());
  });
};