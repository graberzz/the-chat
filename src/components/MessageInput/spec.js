import React from 'react';
import { shallow, mount } from 'enzyme';
import { MessageInput } from './';
import * as api from 'api';

describe('<MessageInput />', () => {
  it('matches snapshot', () => {
    const messageInput = shallow(
      <MessageInput
        api={api}
        roomId='1'
      />
    );

    expect(messageInput).toMatchSnapshot();
  });

  it('matches snapshot with logged in user', () => {
    const user = {
      username: 'user',
      avatar: 'avatar',
      email: 'e@mail.com',
    };

    const messageInput = shallow(
      <MessageInput
        api={api}
        roomId='1'
        user={user}
      />
    );

    expect(messageInput).toMatchSnapshot();
  });

  it('onEmojiClick adds emoji to the textbox', () => {
    const user = {
      username: 'user',
      avatar: 'avatar',
      email: 'e@mail.com',
    };

    const messageInput = mount(
      <MessageInput
        api={api}
        roomId='1'
        user={user}
      />
    );

    messageInput.instance().onEmojiClick('emoji');
    expect(messageInput.state().text).toBe('emoji');
  });
});