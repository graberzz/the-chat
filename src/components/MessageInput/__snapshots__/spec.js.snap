// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageInput /> matches snapshot 1`] = `
<styled.div>
  <TextArea
    disabled={false}
    innerRef={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="You haven't logged in"
    value=""
  />
  <styled.div>
    <EmojiSelector
      onEmojiClick={[Function]}
    />
  </styled.div>
</styled.div>
`;

exports[`<MessageInput /> matches snapshot with logged in user 1`] = `
<styled.div>
  <TextArea
    disabled={false}
    innerRef={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Type your message..."
    value=""
  />
  <styled.div>
    <EmojiSelector
      onEmojiClick={[Function]}
    />
  </styled.div>
</styled.div>
`;
